File: frontend/src/types/analysis.ts
export interface CoverageMetrics {
  total_buildings: number;
  total_households: number;
  total_population: number;
  served_households: number;
  served_population: number;
  underserved_households: number;
  underserved_population: number;
  coverage_percentage: number;
  average_distance: number;
  median_distance: number;
  max_distance: number;
  threshold_meters: number;
  distance_method: string;
  underserved_buildings?: UnderservedBuilding[];
}

export interface UnderservedBuilding {
  building_id: string;
  households: number;
  population: number;
  distance: number;
  center: [number, number];
}

export interface UnderservedCluster {
  cluster_id: string;
  building_count: number;
  households: number;
  population: number;
  center: [number, number];
  avg_distance_to_facility: number;
  priority_score: number;
}

export interface VillageMetrics {
  village_id: string;
  village_name: string;
  total_households: number;
  total_population: number;
  total_buildings: number;
  area_sq_km: number;
  water_facilities: number;
  other_facilities: number;
  water_coverage: CoverageMetrics | null;
  underserved_clusters: UnderservedCluster[];
  priority_level: 'low' | 'medium' | 'high';
  priority_factors: string[];
}

export interface InfrastructureAnalysis {
  infrastructure_type: string;
  facility_count: number;
  coverage: CoverageMetrics;
  underserved_clusters: UnderservedCluster[];
  recommendations: string[];
}

File: backend/app/schemas/analysis.py
"""
Pydantic schemas for analysis endpoints
"""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional


class CoverageMetrics(BaseModel):
    """Coverage analysis metrics"""
    total_buildings: int
    total_households: int
    total_population: int
    served_households: int
    served_population: int
    underserved_households: int
    underserved_population: int
    coverage_percentage: float
    average_distance: float
    median_distance: float
    max_distance: float
    threshold_meters: float
    distance_method: str
    underserved_buildings: List[Dict[str, Any]] = []


class UnderservedCluster(BaseModel):
    """Underserved area cluster"""
    cluster_id: str
    building_count: int
    households: int
    population: int
    center: List[float]
    avg_distance_to_facility: float
    priority_score: float


class VillageMetrics(BaseModel):
    """Complete village metrics"""
    village_id: str
    village_name: str
    
    # Basic stats
    total_households: int
    total_population: int
    total_buildings: int
    area_sq_km: float
    
    # Infrastructure
    water_facilities: int
    other_facilities: int
    
    # Coverage metrics
    water_coverage: Optional[CoverageMetrics] = None
    
    # Underserved areas
    underserved_clusters: List[UnderservedCluster] = []
    
    # Priority assessment
    priority_level: str = "medium"  # low, medium, high
    priority_factors: List[str] = []


class InfrastructureAnalysis(BaseModel):
    """Analysis for specific infrastructure type"""
    infrastructure_type: str
    facility_count: int
    coverage: CoverageMetrics
    underserved_clusters: List[UnderservedCluster]
    recommendations: List[str] = []
